src/transformers/models/dbrx/modeling_dbrx.py

```diff
@@ -0,0 +1,2132 @@
+
+        outputs = (hidden_states,)
+
+        if output_attentions:
+            outputs += (self_attn_weights,)
+
+        if use_cache:
+            outputs += (present_key_value,)
+
+        if output_router_logits:
+            outputs += (router_logits,)
+
+        return outputs
+
+
+class DbrxPreTrainedModel(PreTrainedModel):
+    config_class = DbrxConfig
+    base_model_prefix = 'transformer'
+    supports_gradient_checkpointing = True
+    _no_split_modules = ['DbrxBlock']
+    _skip_keys_device_placement = ['past_key_values']
+    _supports_flash_attn_2 = True
+    _supports_sdpa = False
+    _supports_cache_class = True
+
+    def _init_weights(self, module: nn.Module):
+        std = self.config.initializer_range
+        if isinstance(module, nn.Linear):
+            module.weight.data.normal_(mean=0.0, std=std)
+            if module.bias is not None:
+                module.bias.data.zero_()
+        elif isinstance(module, nn.Embedding):
+            module.weight.data.normal_(mean=0.0, std=std)
+            if module.padding_idx is not None:
+                module.weight.data[module.padding_idx].zero_()
+        elif isinstance(module, nn.LayerNorm):
+            module.weight.data.normal_(mean=0.0, std=std)
+            if module.bias is not None:
+                module.bias.data.zero_()
+        elif isinstance(module, DbrxExpertGLU):
+            module.w1.data.normal_(mean=0.0, std=std)
+            module.v1.data.normal_(mean=0.0, std=std)
+            module.w2.data.normal_(mean=0.0, std=std)
+
+    def _setup_cache(self, cache_cls: Any, max_batch_size: int,
+                     max_cache_len: int):  # TODO: how to set var type of class?
+        if self.config._attn_implementation == 'flash_attention_2' and cache_cls == StaticCache:
+            raise ValueError(
+                '`static` cache implementation is not compatible with ' +
+                '`attn_implementation==flash_attention_2`. Make sure to use ' +
+                '`spda` in the mean time and open an issue at https://github.com/huggingface/transformers.'
+            )
+
+        for block in self.transformer.blocks:
+            device = block.norm_attn_norm.norm_1.weight.device
+            if hasattr(self.config, '_pre_quantization_dtype'):
+                dtype = self.config._pre_quantization_dtype
+            else:
+                dtype = block.norm_attn_norm.attn.out_proj.weight.dtype
+            block.norm_attn_norm.attn.past_key_value = cache_cls(self.config,
+                                                                 max_batch_size,
+                                                                 max_cache_len,
+                                                                 device=device,
+                                                                 dtype=dtype)
+
+    def _reset_cache(self):
+        for block in self.transformer.blocks:
+            block.norm_attn_norm.attn.past_key_value = None
+
+
```

Comment:
let's use copied from as well

Reply:
I think quite a few of the variable names as well as the function def below are different, so I'm not sure if it would work with a "copied from"